### PR TITLE
[CUDA12] Autograd engine use current device only

### PR DESCRIPTION
**Pull request by Aidyn-A**
This is a device agnostic version #91191. 
The reason of existence of this PR is device agnostic policy of autograd engine. Hence, the compile time `USE_CUDA` is not supported, so doing something like:
https://github.com/pytorch/pytorch/blob/fa1ea9f9bcaa77c1370468059be95ad9b421f500/torch/csrc/autograd/engine.cpp#L351-L357
is not effective.

In this PR a check upon CUDA devices in device registry is added such that threads set the same CUDA device.

### CODE REVIEW

**Comment by Aidyn-A** [comment]
Continuing from https://github.com/pytorch/pytorch/pull/94929#issuecomment-1450752448
@albanD this PR is independent on CUDA version and can be tested and benchmarked now.

**Comment by Aidyn-A** [comment]
@albanD do you have any comments on this PR?

**Review by albanD — COMMENTED** [review]
This would deserve a more detailed comment about what it does.
But that only helps if you already initialized cuda on at least one device?

If you're doing such a disruptive change, I guess at this point you might as well just move the set_device into the inner loop after we get work: https://github.com/pytorch/pytorch/blob/b0b5f3c6c681896febbd9ff7ad7649b13def345d/torch/csrc/autograd/engine.cpp#L516
And make sure that when we don't change the device, this will be super cheap to do ?
cc @ngimel

File: torch/csrc/autograd/engine.cpp [hunk 1]
```diff
@@ -783,12 +777,19 @@ void set_device(int device) {
   //
   // Don't use DeviceGuard here because its destructor may be called before the
   // device is reset. This is fine because the device is thread local.
+  //
+  // Setting the CUDA device is being omitted because it will unnecessarily
```
[the remark this answers is not in the document]
Comment (Aidyn-A) [reply torch/csrc/autograd/engine.cpp:781]: Yes, I thought of the simplest change code path. Please, correct me if I am wrong. The device behavior of the device guard would be something like this:
```Python
torch.cuda.set_device(0)
a = torch.ones(10, device="cuda:1", requires_grad=True) # set cuda:1 -> allocate the memory  -> set cuda:0
b = torch.sum(a) # set cuda:1 -> computes sum  -> set cuda:0
b.backward() # set cuda:1 -> call autograd engine -> set cuda:0
```
The why would we want to set anything in the engine if the required device is already set. 
(If at the beginning `torch.cuda.set_device(1)` then device guard is no-op)

[the remark this answers is not in the document]
Comment (albanD) [reply torch/csrc/autograd/engine.cpp]: There are multiple threads involved in the autograd engine. So each needs to do the right thing.
The problem here is that without a global guard, every single op within the backward will set the device to 1 then back to 0.
While the current code let the guard bypass setting the device altogether as it is already 1 to begin with.

File: torch/csrc/autograd/engine.cpp [hunk 2]
```diff
@@ -526,6 +519,8 @@ auto Engine::thread_main(const std::shared_ptr<GraphTask>& graph_task) -> void {
         continue;
       }
 
+      set_device(worker_device);
```
Comment (albanD) [inline torch/csrc/autograd/engine.cpp:522]: This function is relatively expensive. I think we should have a version that:
- skip the tls set
- Only creates the Device and device guards if the device is not properly set already. Checking the current device should be a lot cheaper than all of that right?

**Review by albanD — COMMENTED** [review]
Thanks for the update the change is at the right place, only a small perf concern

File: torch/csrc/autograd/engine.cpp (hunk 2 above)
Comment (Aidyn-A) [reply torch/csrc/autograd/engine.cpp:522]: Thanks for the response.
- Skipping tls guard https://github.com/pytorch/pytorch/blob/1e6961586b9d213e05cd5ecbd5a1c17acbcb051d/torch/csrc/autograd/engine.cpp#L533 caused test failures (ran locally). 

- device guard skips setDevice if it is already set: https://github.com/pytorch/pytorch/blob/1e6961586b9d213e05cd5ecbd5a1c17acbcb051d/c10/cuda/impl/CUDAGuardImpl.h#L52-L58
despite that I attempted to skip creating device guard for CUDA in [df53cb8](https://github.com/pytorch/pytorch/pull/92354/commits/df53cb8e0a19e20c6685c2abe111b68be8323dec)

**Review by albanD — APPROVED** [review]
Thanks!

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**